### PR TITLE
Licensing, Style, and Structure Code Review

### PR DESCRIPTION
**Pull request by jercytryn**

### CODE REVIEW

File: build/lib/pdbvangogh/__init__.py
Comment (jercytryn) [inline build/lib/pdbvangogh/__init__.py:1]: Can this dir be excluded from the repo? It's just a build artifact right? Consider adding /build/ to a [.gitignore](https://github.com/github/gitignore/blob/main/Python.gitignore) [file](https://git-scm.com/docs/gitignore)

File: src/tests.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+                   pdb_id = '2HYY',
+                   out_dir = 'tests/out/',
+                   style_image = 'tests/in/starry_night.png',
+                   save_prefix = 'tests/out/2HYY_pdx_starry_night',
+                   content_size=100,
+                   background_size = 200,
+                   background_hyperparameters=vgg19_transfer_parameters(style_weight=1e-2, epochs=1, steps_per_epoch=50),
+                   content_hyperparameters=vgg19_transfer_parameters(style_weight=1e-4, epochs=1, steps_per_epoch=50)
+                   )
+    assert(True)
```
Comment (jercytryn) [inline src/tests.py:37]: Nice, some tests :) I'm guessing you just do a visual or manual diff inspection for now? Though may be worth just adding a couple more basic asserts for now that like the image is output to the path that you expect, etc. Or you can even use something like [imgdiff](https://pypi.org/project/imgdiff/#:~:text=A%20command%2Dline%20tool%20that,a%20version%20control%20tool%2C%20e.g.) if you want to make it entirely procedural 

Another fun thing you can add later is to use github actions to automatically run your tests e.g. on pull request, etc.

File: build/lib/pdbvangogh/models.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    steps_per_epoch = hyperparameters.steps_per_epoch
+    opt = tf.keras.optimizers.Adam(learning_rate=0.02, beta_1=0.99, epsilon=1e-1)
+    step = 0
+    in_jupyter_notebook = am_i_in_a_jupyter_notebook()
+    for n in range(epochs):
+        for m in range(steps_per_epoch):
+            step += 1
+            train_step(image, style_targets, content_targets)
+            print(".", end="", flush=True)
+        # if in_jupyter_notebook:
```
Comment (jercytryn) [inline build/lib/pdbvangogh/models.py:129]: nit: clean this up or uncomment?  Since you already go through the pains of knowing whether you are in an interactive context I'm guessing should be safe?

File: src/pdbvangogh/models.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    style_layers = ["block1_conv1", "block2_conv1", "block3_conv1", "block4_conv1", "block5_conv1"]
+    num_content_layers = len(content_layers)
+    num_style_layers = len(style_layers)
+    style_extractor = vgg_layers(style_layers)
+    style_outputs = style_extractor(style_image * 255)
+    extractor = StyleContentModel(style_layers, content_layers)
+    results = extractor(tf.constant(content_image))
+    style_targets = extractor(style_image)["style"]
+    content_targets = extractor(content_image)["content"]
+    start = time.time()
```
Comment (jercytryn) [inline src/pdbvangogh/models.py:104]: you're really only doing this once so not a big thing, but consider using a timing decorator or context manager if you start doing more timing / time measuring/reporting operations.  Also the advantage of the decoupling the implementation of the timing and display from the function itself

File: src/pdbvangogh/img.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    if len(image.shape) > 3:
+        image = tf.squeeze(image, axis=0)
+
+    plt.imshow(image)
+    if title:
+        plt.title(title)
+
+
+def resize_pil(image, new_width):
+    from PIL import Image
```
Comment (jercytryn) [inline src/pdbvangogh/img.py:42]: style nit: in general imports should go at the top of the file, if it's needed inline it usually hints at a circular dependency in the code that should be addressed.  In this case though, import not needed?

File: src/pdbvangogh/img.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+    original_width, original_height = image.size
+    aspect_ratio = original_height / original_width
+    new_height = int(new_width * aspect_ratio)
+    resized_image = image.resize((new_width, new_height))
+    return resized_image
+
+
+def overlay_images(background_path, foreground_path, position=(0, 0), background_size=None, foreground_size=None, save_path="overlayed_image.png"):
+    from PIL import Image
```
Comment (jercytryn) [inline src/pdbvangogh/img.py:52]: ditto - move import to top of file

File: src/pdbvangogh/data_models.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+class structure(BaseModel):
+    id: str = None  # the pdb ID of the structure of interest
+    cif: str = None  # the cif file of the
+    width: int = 800  # the width of the image made from the cif file
+    height: int = 600  # the height of the image made from the cif file
+    size: int = 500  # the size of the larger
+
+
+class background(BaseModel):
```
Comment (jercytryn) [inline src/pdbvangogh/data_models.py:14]: worth making this inherit from like a general image model?

File: src/pdbvangogh/data_models.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    image_path: str  # the path of the background image
+    size: int = None  # the size of the image
+
+
+class style(BaseModel):
```
Comment (jercytryn) [inline src/pdbvangogh/data_models.py:19]: could style and background use the same model since size is optional anyways?

File: src/pdbvangogh/api.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    save_prefix,
+    out_dir=None,
+    content_image=None,
+    pdb_id=None,
+    background_size = 500,
+    content_size = 800, 
+    background_hyperparameters=vgg19_transfer_parameters(style_weight=1e-2, epochs=10, steps_per_epoch=100),
+    content_hyperparameters=vgg19_transfer_parameters(style_weight=1e-4, epochs=10, steps_per_epoch=100),
+    debug=False,
+):
```
Comment (jercytryn) [inline src/pdbvangogh/api.py:24]: you may want to add a docstring here if this is the top-level api describing the parameters, etc..  google style or sphinx are popular formats

File: src/pdbvangogh/api.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+from pdbvangogh import pdb
+from pdbvangogh.img import *
+from pdbvangogh.models import *
+from pdbvangogh.data_models import *
+import os
+
+
+def test_func():
```
Comment (jercytryn) [inline src/pdbvangogh/api.py:8]: is this just a placeholder for now?

File: README.md [hunk 10]
```diff
@@ -0,0 +1 @@
+pdbvangogh is a package intended to enable style transfer on molecular structures with artistic backgrounds
```
Comment (jercytryn) [inline README.md:1]: Typically you'd add a top-level LICENSE file at the root level w/whatever you end up going with (MIT, Apache 2.0, etc.)

File: src/pdbvangogh/__init__.py
Comment (jercytryn) [inline src/pdbvangogh/__init__.py:1]: Small suggestion: considering importing the top level api function to this level.  More convenient for consumers of the api to just be able to access the main function from the top level of your lib rather than needing to know there's an api submodule.  Has the benefit of "hiding" internals from end client user, assuming they don't need to leverage them

ie
```
from .api import pdbvangogh
```

File: src/pdbvangogh/data_models.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    image_path: str  # the path of the style image
```
Comment (jercytryn) [inline src/pdbvangogh/data_models.py:20]: are these currently used anywhere?

**Review by jercytryn — COMMENTED** [review]
Looks good to me! Clear, concise, seems to be pretty well modularized, easy to follow, etc.  Could go a more object-oriented route but given the code complexity rn it would probably be overkill anyways.  Docstrings would be a nice to have, at least at the top of each module and for the main api function.  Well done on getting some tests in there from the get go as well.  

You may want to consider using proper [python logging](https://docs.python.org/3/library/logging.html) instead of print statements so that consumers of the api can more easily configure / turn on off print statements and different levels (to stderr etc. ) ping me if you want a quick overview on that

I left some specific inline comments / suggestions, take them or leave them.  

wrt the tensorflow code, it seems like it was fairly minimal what was used from there but if a significant portion is line-for-line, I would just also include the tensorflow license somewhere in here (even under source and mention that somewhere in your code in a comment).  Otherwise, I think you are good there, you are not distributing any third party code so Apache 2.0 and MIT, etc. of those packages are satisfied :)

Other than that, really cool project with some awesome results, well done 🚀
